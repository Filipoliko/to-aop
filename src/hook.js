export function createHook(name, regular, callback) {
  return {
    [name]: function hook(meta) {
      return hookFor(meta, regular, callback);
    }
  };
}

export function hookFor(meta, filterRule, callback) {
  if (typeof filterRule === 'string') {
    if (meta.property.includes(filterRule)) {
      return callback(meta);
    }

    return null;
  }
  if (filterRule instanceof RegExp) {
    if (filterRule.test(meta.property)) {
      return callback(meta);
    }

    return null;
  }

  if (typeof filterRule === 'function') {
    if (filterRule(meta)) {
      return callback(meta);
    }

    return null;
  }

  if (Array.isArray(filterRule)) {
    return filterRule.map(({ rule, action }) => {
      return hookFor(meta, rule, action || callback);
    });
  }

  throw new TypeError(
    `Invalid rule type ${typeof filterRule}. Method accept string, regexp, function and array.`
  );
}

export const hookName = Object.freeze({
  beforeMethod: 'beforeMethod',
  afterMethod: 'afterMethod',
  aroundMethod: 'aroundMethod',
  beforeGetter: 'beforeGetter',
  afterGetter: 'afterGetter',
  aroundGetter: 'aroundGetter',
  beforeSetter: 'beforeSetter',
  afterSetter: 'afterSetter',
  aroundSetter: 'aroundSetter'
});
