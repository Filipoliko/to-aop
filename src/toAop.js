'use strict';

import { createProxy, createCallTrap } from './trap';
import { AOP_PATTERN, AOP_HOOKS } from './symbol';

export function createAspect(pattern) {
  return function applyAop(target) {
    return aop(target, pattern);
  };
}

export function aop(target, pattern) {
  if (Object.hasOwnProperty(target, AOP_PATTERN)) {
    mergePattern(target, pattern);

    return;
  }

  Reflect.defineProperty(target, AOP_PATTERN, {
    value: pattern,
    enumerable: false,
    writable: true
  });

  if (typeof target === 'function') {
    return applyAopToClass(target);
  }

  if (typeof target === 'object') {
    return applyAopToInstance(target);
  }

  throw new TypeError(
    `aop accept only object and class. You gave type of ${typeof target}.`
  );
}

function applyAopToInstance(instance) {
  return createProxy(instance);
}

function applyAopToClass(target) {
  let pattern = target[AOP_PATTERN];

  applyAopToStaticMethods(target, pattern);
  applyAopToMethods(target, pattern);
}

function applyAopToStaticMethods(target, pattern) {
  let original = {};

  while (target && target !== Function.prototype) {
    Object.entries(Object.getOwnPropertyDescriptors(target)).forEach(
      ([property]) => {
        if (typeof target[property] === 'function') {
          original[property] = target[property];

          target[property] = createCallTrap(
            target,
            original,
            property,
            pattern
          );
        }
      }
    );

    target = Reflect.getPrototypeOf(target);
  }
}

function applyAopToMethods(target, pattern) {
  let original = {};
  let prototype = target.prototype;
  while (prototype) {
    Object.entries(Object.getOwnPropertyDescriptors(prototype)).forEach(
      function([property]) {
        try {
          if (property in original) {
            return;
          }
          original[property] = prototype[property];

          // TODO override getter and setter

          let aopHooks = original[property][AOP_HOOKS];
          if (aopHooks) {
            const { object } = aopHooks[aopHooks.length - 1];
            aopHooks.push({
              target,
              object,
              property,
              pattern
            });
            return;
          }

          if (typeof original[property] === 'function') {
            prototype[property] = createCallTrap(
              target,
              original,
              property,
              pattern
            );
          }
        } catch (_) {
          console.error(_);
        } // eslint-disable-line no-empty
      }
    );

    prototype = Reflect.getPrototypeOf(prototype);
  }
}

function mergePattern(target, pattern) {
  let currentTargetPattern = target[AOP_PATTERN];

  target[AOP_PATTERN] = Object.entries(pattern).reduce(
    (resultPattern, [hookName, hookValue]) => {
      if (!resultPattern[hookName]) {
        resultPattern = hookName;
      }

      if (resultPattern[hookName]) {
        if (!Array.isArray(resultPattern[hookName])) {
          resultPattern[hookName] = [resultPattern[hookName]];
        }
        if (!Array.isArray(hookValue)) {
          hookValue = [hookValue];
        }

        resultPattern[hookName] = resultPattern[hookName].concat(hookValue);
      }

      return resultPattern;
    },
    currentTargetPattern
  );
}
