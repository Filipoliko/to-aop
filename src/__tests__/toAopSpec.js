import { createAspect } from '../toAop';
import { hookName } from '../hook';
import createClasses from './createClasses';
import createPattern from './createPattern';

describe('createAspect method', () => {
  let withAspect = null;
  let pattern = null;

  beforeEach(() => {
    pattern = createPattern();

    withAspect = createAspect(pattern);
  });

  describe('for instance', () => {
    let beforeMethod = null;
    let beforeGetter = null;
    let afterMethod = null;

    beforeEach(() => {
      beforeMethod = jest.fn();
      beforeGetter = jest.fn();
      afterMethod = jest.fn();

      pattern = createPattern(undefined, {
        [hookName.beforeMethod]: beforeMethod,
        [hookName.beforeGetter]: beforeGetter,
        [hookName.afterMethod]: afterMethod
      });

      withAspect = createAspect(pattern);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod', () => {
      let { A } = createClasses();
      let instance = withAspect(new A());
      let result = instance.method({}, 1);

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(beforeMethod.mock.calls[0]).toMatchSnapshot();
      expect(afterMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls[0]).toMatchSnapshot();

      expect(result).toEqual(undefined);
    });

    it('should call pattern.beforeGetter and pattern.beforeMethod', () => {
      let { C } = createClasses();
      let instance = withAspect(new C());
      let entries = instance.map.entries();

      expect(beforeGetter.mock.calls.length).toEqual(2);
      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(entries.next().done).toEqual(true);
    });
  });

  describe('for object', () => {
    let beforeMethod = null;
    let afterMethod = null;

    beforeEach(() => {
      beforeMethod = jest.fn();
      afterMethod = jest.fn();

      pattern = createPattern(undefined, {
        [hookName.beforeMethod]: beforeMethod,
        [hookName.afterMethod]: afterMethod
      });

      withAspect = createAspect(pattern);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for dummy object', () => {
      let dummy = {
        method() {}
      };

      let xdummy = withAspect(dummy);
      xdummy.method('works');

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls.length).toEqual(1);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for global console', () => {
      spyOn(console, 'log');

      let xconsole = withAspect(console);
      xconsole.log('works');

      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls.length).toEqual(1);
    });

    it('should call pattern.beforeMethod and pattern.afterMethod for object with context', () => {
      const object = {
        value: 1,
        method: function increase() {
          this.value++;
        }
      };

      const xobject = withAspect(object);

      xobject.method();

      expect(xobject.value).toEqual(2);
      expect(beforeMethod.mock.calls.length).toEqual(1);
      expect(afterMethod.mock.calls.length).toEqual(1);
    });
  });
});
