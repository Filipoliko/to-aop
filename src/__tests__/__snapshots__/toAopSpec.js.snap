// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod 1`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": A {
      "variable": "method",
    },
    "object": Object {
      "method": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod 2`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": A {
      "variable": "method",
    },
    "object": Object {
      "method": [Function],
    },
    "payload": "method",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod after class is instanced 1`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "object": Object {
      "method": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod after class is instanced 2`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "object": Object {
      "method": [Function],
    },
    "payload": "B method",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod for extended class 1`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "object": Object {
      "method": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod for extended class 2`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "object": Object {
      "method": [Function],
    },
    "payload": "B method",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod for extended classes with same ancestor 1`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "object": Object {
      "method": [Function],
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod for extended classes with same ancestor 2`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": B {
      "variable": "method",
    },
    "object": Object {
      "method": [Function],
    },
    "payload": "method",
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod for static method 1`] = `
Array [
  Object {
    "args": Array [],
    "context": [Function],
    "object": Object {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "property": "staticMethod",
    "target": [Function],
  },
]
`;

exports[`createAspect method for class should call pattern.beforeMethod and pattern.afterMethod for static method 2`] = `
Array [
  Object {
    "args": Array [],
    "context": [Function],
    "object": Object {
      "staticMethod": [Function],
      "staticMethod2": [Function],
    },
    "payload": "static method",
    "property": "staticMethod",
    "target": [Function],
  },
]
`;

exports[`createAspect method for instance should call pattern.beforeMethod and pattern.afterMethod 1`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": A {
      "variable": undefined,
      Symbol(AopPattern): Object {
        "afterGetter": [Function],
        "afterMethod": [Function],
        "beforeGetter": [Function],
        "beforeMethod": [Function],
      },
    },
    "object": A {
      "variable": undefined,
      Symbol(AopPattern): Object {
        "afterGetter": [Function],
        "afterMethod": [Function],
        "beforeGetter": [Function],
        "beforeMethod": [Function],
      },
    },
    "property": "method",
    "target": [Function],
  },
]
`;

exports[`createAspect method for instance should call pattern.beforeMethod and pattern.afterMethod 2`] = `
Array [
  Object {
    "args": Array [
      Object {},
      1,
    ],
    "context": A {
      "variable": undefined,
      Symbol(AopPattern): Object {
        "afterGetter": [Function],
        "afterMethod": [Function],
        "beforeGetter": [Function],
        "beforeMethod": [Function],
      },
    },
    "object": A {
      "variable": undefined,
      Symbol(AopPattern): Object {
        "afterGetter": [Function],
        "afterMethod": [Function],
        "beforeGetter": [Function],
        "beforeMethod": [Function],
      },
    },
    "payload": undefined,
    "property": "method",
    "target": [Function],
  },
]
`;
